design/baremetal-operator/bmh-v1beta1.md

```diff
@@ -0,0 +1,204 @@
+
+  Remove `RootGB` (only makes sense for partition images which we don't
+  support) and `LocalGB` (optional and discovered by inspection).
+
+- Prepare `Image.ChecksumType` for the default change:
+
+  Add a new type `auto` that will tell Ironic to detect the checksum type
+  itself (by not providing the `image_os_hash_algo` field).
+
+- Ensure that a `HardwareData` CR exists for any host that has
+  `HardwareDetails` in the status.
+
+The following significant changes will be done to the new version:
+
+- Check the default value of `Image.ChecksumType` from `md5` to `auto`.
+  Mark the `md5` type as deprecated for removal before `v1`.
+
+- Remove `HardwareProfile`, effectively using the `empty` profile in the
+  new API version.
+
+- Remove `HardwareDetails` from the `Status` in favour of the already
+  introduced `HardwareData` CR.
+
+The following cosmetic changes will be done to the new version:
+
+- Rename `online` to `poweredOn`.
+
+  The word *online* is ambiguous here: it may be understood to refer to
+  provisioning status. Furthermore, the corresponding field in the `Status`
+  is already called `poweredOn`.
+
+- In `RootDeviceHints`:
+
+  Rename `DeviceName` to `DevicePath` to better reflect what it actually is
+  (especially after introducing support for `/dev/disk/by-path`).
+
+  Rename `Model` and `Vendor` to `ModelContains` and `VendorContains` to
+  reflect the fact that these fields use the `<in>` operator.
+
+### Implementation Details/Notes/Constraints
+
+Since the new API version will mostly involve removing or renaming fields,
+the storage version will stay `v1alpha1` until this version is removed.
+Similarly, the controller will also the old version to be able to handle
+the fields not present in the new one (and to avoid constant conversion to
+and from the storage version).
+
+### Risks and Mitigations
+
+None
+
+### Work Items
+
+- Refactor the BMO code to avoid referring directly to the current version.
+- Add `Architecture` to the current version.
+- Add `empty` hardware profile.
+- Add `auto` checksum type.
+- Copy type definitions and modify them.
+- Create a conversion webhook to change between versions.
+- Make a new BMO release and communicate the change to the community.
+
+The old version will be phased out as follows:
+
+- Create the last BMO branch with `v1alpha1` fully functional.
+
+- Switch the storage and controller client versions to `v1beta1`. At this
+  point, `v1alpha1` will still be generally functional, except for the features
+  that are not part of `v1beta1`
+
+- Create a migration script that reads `v1alpha1` resources and converts them
+  to `v1beta1` ones. This is to ensure that each resource is written at least
+  once with the new storage version.
+
+- Create a new BMO release to get this change to the operators. Instruct them
+  to run the migration script or migrate the objects themselves.
+
+- Delete `v1alpha1` from the supported versions, remove the conversion webhook.
```

Comment:
I'm not clear on why we need to do this? We could leave that code ~forever without it hurting anything, so why force a rapid change that requires a manual migration within 3-6 months?

Reply:
@Rozzii @kashifest @lentzi90 what's your opinion on this thread?

Reply:
The above looks alright to me. I interpret it as just describing how the old version will be phased out (not when).
However, line 135 below changes that a bit. Is the idea to phase out and remove `v1alpha1` in 3-6 months after `v1beta1`? That sounds like a bad idea to me.

I would expect at least a couple of releases in between with both API versions supported. Commonly it has been done so that the old version is first marked as deprecated (but still fully functional) to give a heads up to users. Then in some later release there is a big warning and we stop serving it. At this point, it is still possible to enable it again just through `served: true/false` in the CRD, so ignorant users will get a big scare, but can easily be "saved". Then in the next release the old API version would be completely removed from the code base.

Reply:
A good example of how we plan to remove v1a5 api from CAPM3 is presented here https://github.com/metal3-io/cluster-api-provider-metal3/issues/971. I agree with @dtantsur that we can decide this as a separate issue/discussion.